polynote-runtime/src/main/scala/polynote/runtime/DataEncoder.scala

```diff
@@ -0,0 +1,163 @@
+  implicit val int: DataEncoder[Int] = instance(IntType)(_ writeInt _)
+  implicit val long: DataEncoder[Long] = instance(LongType)(_ writeLong _)
+  implicit val float: DataEncoder[Float] = instance(FloatType)(_ writeFloat _)
+  implicit val double: DataEncoder[Double] = instance(DoubleType)(_ writeDouble _)
+  implicit val string: DataEncoder[String] = instance(StringType) {
+    (out, str) =>
+      out.writeInt(str.length)
+      out.write(str.getBytes(StandardCharsets.UTF_8))
+  }
+
+  implicit val byteArray: DataEncoder[Array[Byte]] = instance(BinaryType) {
+    (out, bytes) =>
+      out.writeInt(bytes.length)
+      out.write(bytes)
+  }
+
+  implicit val byteBuffer: DataEncoder[ByteBuffer] = instance(BinaryType) {
+    (out, bytes) =>
+      val b = bytes.duplicate()
+      b.rewind()
+      out.writeInt(b.limit())
+      val buf = new Array[Byte](1024)
+      while (b.hasRemaining) {
+        val len = math.min(1024, b.remaining())
+        b.get(buf, 0, len)
+        out.write(buf, 0, len)
+      }
+  }
+
+  implicit def array[A](implicit encodeA: DataEncoder[A]): DataEncoder[Array[A]] = sizedInstance[Array[A]](ArrayType(encodeA.dataType), arr => arr.length * encodeA.dataType.size) {
+    (output, arr) =>
+      output.writeInt(arr.length)
+      var i = 0
+      while (i < arr.length) {
+        encodeA.encode(output, arr(i))
+        i += 1
+      }
+  }
+
+  implicit def traversable[F[X] <: GenTraversable[X], A](implicit encodeA: DataEncoder[A]): DataEncoder[F[A]] = sizedInstance[F[A]](ArrayType(encodeA.dataType), seq => seq.size * encodeA.dataType.size) {
+    (output, seq) =>
+      output.writeInt(seq.size)
+      seq.foreach(encodeA.encode(output, _))
+  }
+
+  private[polynote] class BufferOutput(buf: ByteBuffer) extends DataOutput {
+    def write(b: Int): Unit = buf.put(b.toByte)
+    def write(b: Array[Byte]): Unit = buf.put(b)
+    def write(b: Array[Byte], off: Int, len: Int): Unit = buf.put(b, off, len)
+    def writeBoolean(v: Boolean): Unit = buf.put(if (v) 1.toByte else 0.toByte)
+    def writeByte(v: Int): Unit = buf.put(v.toByte)
+    def writeShort(v: Int): Unit = buf.putShort(v.toShort)
+    def writeChar(v: Int): Unit = buf.putChar(v.toChar)
+    def writeInt(v: Int): Unit = buf.putInt(v)
+    def writeLong(v: Long): Unit = buf.putLong(v)
+    def writeFloat(v: Float): Unit = buf.putFloat(v)
+    def writeDouble(v: Double): Unit = buf.putDouble(v)
+    def writeBytes(s: String): Unit = buf.put(s.toCharArray.map(_.toByte))
+    def writeChars(s: String): Unit = s.toCharArray.foreach(c => writeChar(c))
+    def writeUTF(s: String): Unit = {
+      // the logic for this is actually quite complex... it's all in static function [[java.io.DataOutputStream.writeUTF]], but that's package private.
+      // Nobody ought to be using this anyway, though, because it's not how we write strings. So we'll just do it inefficiently by going through a byte array.
```

Comment:
Doing it this way isn't a problem but just curious - if we don't expect this to ever be called why not throw a NotImplementedError? At least that way we'd know if it ever did get called.